marlowe-dashboard-client/src/Capability/PlutusApps/MarloweApp.purs

```diff
@@ -38,48 +47,134 @@ import Plutus.V1.Ledger.Value (TokenName) as Back
+  PlutusAppId ->
+  UUID ->
+  String ->
+  Lens' EndpointMutex (AVar Unit) ->
+  payload ->
+  m (AjaxResponse Unit)
+invokeMutexedEndpoint plutusAppId reqId endpointName _endpointMutex payload = do
+  -- There are three mutex involved in this operation:
+  --   * The endpointMutex help us avoid making multiple requests to the same endpoint if its not
+  --     available.
+  --   * The global requestMutex help us manage the request queue concurrently.
+  --   * For each request in the queue, we store one mutex that represents the response.
+  endpointMutex <- asks $ view (_marloweAppEndpointMutex <<< _endpointMutex)
+  requestMutex <- asks $ view (_marloweAppEndpointMutex <<< _requests)
+  -- First we need to see if the endpoint is available to make a request
+  -- TODO: we could later add a forkAff with a timer that unlocks this timer if we
+  --       dont get a response
+  -- TODO: We could change this take for a read and listen for a 500 error with EndpointUnavailabe
+  --       to retry and take it (instead of preemptively taking it).
+  liftAff $ AVar.take endpointMutex
+  result <- Contract.invokeEndpoint plutusAppId endpointName payload
+  -- After making the request, we lock on the global request array so we can add
+  -- a new request id with its corresponding mutex.
+  -- TODO: It would be nice if the invokeEndpoint returns a requestId instead of having to generate
+  --       one from the FE.
+  liftAff do
+    requests <- AVar.take requestMutex
+    newRequestMutex <- AVar.empty
+    let
+      -- We add new requests to the begining of the array and remove them from the end.
+      newRequests = take 15 $ (reqId /\ newRequestMutex) : requests
+    AVar.put newRequests requestMutex
+  pure result
+
+-- When a MarloweApp emits a new observable state, it comes with the information of which
+-- request id originated the change. We use this function to update the request queue and
+-- free any thread that might be waiting for the result of an action.
+-- The return type is a Maybe of the same input, if the value is Nothing, then we couldn't
+-- find a request that triggered this response. This can happen for example when reloading
+-- the webpage, or if we have two browsers with the same wallet.
+onNewObservableState ::
+  forall env m.
+  MonadAff m =>
+  MonadAsk (MarloweAppEndpointMutexEnv env) m =>
+  LastResult ->
+  m (Maybe LastResult)
+onNewObservableState lastResult = case lastResult of
+  OK reqId _ -> onNewObservableState' reqId
+  SomeError reqId _ _ -> onNewObservableState' reqId
+  _ -> pure Nothing
+  where
+  onNewObservableState' reqId = do
+    requestMutex <- asks $ view (_marloweAppEndpointMutex <<< _requests)
+    -- This read is blocking but does not take the mutex.
+    requests <- liftAff $ AVar.read requestMutex
+    case findReqId reqId requests of
+      Nothing -> pure Nothing
+      Just reqMutex -> do
+        liftAff $ AVar.put lastResult reqMutex
+        pure $ Just lastResult
```

Comment:
can replace pattern match with `for` / `traverse`

```suggestion
    for (findReqId reqId requests) \reqMutex -> do
      liftAff $ AVar.put lastResult reqMutex
      pure lastResult
```

Reply:
👍🏻